scripts/anvil_tools/add_or_remove_user_from_project.py

```diff
@@ -0,0 +1,111 @@
+from pandas import *
+import requests
+import urllib.parse
+from oauth2client.client import GoogleCredentials
+
+# function to get authorization bearer token for requests
+def get_access_token():
+    """Get access token."""
+
+    scopes = ["https://www.googleapis.com/auth/userinfo.profile", "https://www.googleapis.com/auth/userinfo.email"]
+    credentials = GoogleCredentials.get_application_default()
+    credentials = credentials.create_scoped(scopes)
+    return credentials.get_access_token().access_token
+
+
+def add_or_remove_user_from_project(project, email_list, add_action, verbose):
+    """Adding or Removing list of users' emails to the billing project"""
+    if verbose:
+        print(email_list)
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: */*" -H  "Authorization: Bearer [token] -H "Content-Type: application/json"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "*/*", "Content-Type": "application/json"}
+    
+    # Stopping running code at the 6th error to avoid multiple failures
+    error_count = 0
+
+    # Keeping track of emails added
+    emails_added = 0
+
+    for email in email_list:
+        # URL Encoding the email
+        url_encode_email = urllib.parse.quote(email)
+        
+        # Library/putLibraryMetadata
+        uri = f"https://rawls.dsde-prod.broadinstitute.org/api/billing/v2/{project}/members/user/{url_encode_email}"
+        
+        # capture response from API
+        if add_action:
+            response = requests.put(uri, headers=headers)
+            action="add"
+            action_past="added"
+        else:
+            response = requests.delete(uri, headers=headers)
+            action="delete"
+            action_past="deleted"
+
+        # Getting status code
+        status_code = response.status_code
+        
+        # adding fail message
+        if status_code != 200:
+            error_count+=1
+            print(f"WARNING: Failed to {action} the email {email} to the billing project {project}.")
+            print("Please see full response for error:")
+            print(response.text)
+             
+            if error_count > 6:
+                print("ERROR: Failed to run on 6 emails")
+                return "ERROR: Failed to run on 6 emails"
+        else:
+            # adding success message
+            emails_added+=1
+
+            # Print verbose message
+            if verbose:
+                print(f"Successfully {action_past} the email {email} to the billing project {project}.")
+    
+    # Success of Fail Messages
+    if error_count < 1:
+        print(f"Successfully {action_past} all emails to the billing project: {project}.")
+    elif error_count == len(email_list):
+        print(f"Failed to {action} emails to the billing project: {project}.")
+    else:
+        print(f"Successfully {action_past} only {emails_added} emails out of {len(email_list)} to the billing project: {project}.")
```

Comment:
as I think about this more, I think also printing out the list of emails that failed to {action} would be useful to the user. I know they're printed out as they go, but I worry the error messages might overwhelm the user from pulling out the emails that failed.

Reply:
Sure, I will add it now